src/litserve/docker_builder.py

```diff
@@ -0,0 +1,217 @@
+        # Create a tar archive of the temporary directory
+        try:
+            tar_stream = create_tar_stream(tmpdir)
+        except Exception as e:
+            logger.error(f"Failed to create tar archive: {e}")
+            return None
+
+        # Build the Docker image using the low-level API
+        try:
+            logger.info(f"Starting build for image '{tag}'...")
+            build_logs = client.build(
+                fileobj=tar_stream, custom_context=True, tag=tag, rm=True, decode=True, timeout=timeout
+            )
+            image_id = None
+            for chunk in build_logs:
+                if "stream" in chunk:
+                    log_message = chunk["stream"].strip()
+                    if log_message:
+                        logger.info(log_message)
+                elif "error" in chunk:
+                    logger.error(chunk["error"].strip())
+            logger.info(f"Image '{tag}' built successfully.")
+            # Optionally, retrieve the image ID
+            images = client.images(name=tag)
+            if images:
+                image_id = images[0]["Id"]
+                logger.debug(f"Image ID: {image_id}")
+            return image_id
+        except docker.errors.BuildError as e:
+            logger.error(f"Build failed: {e}")
+            return None
+        except docker.errors.APIError as e:
+            logger.error(f"Docker API error during build: {e}")
+            return None
+        finally:
+            tar_stream.close()
+
+
+def create_tar_stream(directory: str) -> io.BytesIO:
+    """Creates a gzip-compressed tar archive from the specified directory.
+
+    Args:
+        directory (str): The path to the directory to archive.
+
+    Returns:
+        io.BytesIO: An in-memory tar archive.
+
+    """
+    import io
+
+    tar_stream = io.BytesIO()
+    with tarfile.open(fileobj=tar_stream, mode="w:gz") as tar:
+        tar.add(directory, arcname=".")
+    tar_stream.seek(0)
+    return tar_stream
+
+
+def build(server_path: str, tag: Optional[str] = None, port: int = 8000, timeout: int = 600):
+    """Build a Docker image from the given server code.
+
+    Args:
+        server_path (str): The path to the server file.
+        tag (str, optional): The tag of the Docker image. Defaults to 'litserve:latest'.
+        port (int, optional): The port to expose. Defaults to 8000.
+        timeout (int, optional): The timeout for building the Docker image in seconds. Defaults to 600.
+
+    """
+    if not tag:
+        tag = "litserve:latest"
+
+    client = get_docker_client()
+
+    files = []
+    requirements = ""
+    if os.path.exists(REQUIREMENTS_FILE):
+        requirements = f"-r {REQUIREMENTS_FILE}"
+        files.append(REQUIREMENTS_FILE)
+    else:
+        warnings.warn(
+            f"requirements.txt not found at {os.getcwd()}. "
+            f"Make sure to install the required packages in the Dockerfile."
+        )
+    files.append(server_path)  # TODO: Make it flexible
```

Comment:
This is rather limiting, we need to add a path (a directory) to include in the build.